File: backend/core-api/src/test/java/ai/govbiz/core/service/AiServiceHealthServiceTest.java
package ai.govbiz.core.service;

import java.net.ConnectException;
import java.net.SocketTimeoutException;
import java.util.stream.Stream;

import ai.govbiz.core.client.ai.AiServiceClient;
import ai.govbiz.core.client.ai.AiServiceClientException;
import ai.govbiz.core.client.ai.AiServiceHealthPayload;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class AiServiceHealthServiceTest {

    @Mock
    private AiServiceClient client;

    @InjectMocks
    private AiServiceHealthService service;

    @Test
    void mapsValidatedPayloadToApplicationResult() {
        when(client.getHealth()).thenReturn(
                new AiServiceHealthPayload("up", "govbiz-ai-service"));

        AiServiceHealthResult result = service.getHealth();

        assertEquals("up", result.status());
        assertEquals("govbiz-ai-service", result.service());
    }

    @ParameterizedTest
    @MethodSource("invalidPayloads")
    void rejectsMissingOrUnexpectedLiteralValues(AiServiceHealthPayload payload) {
        when(client.getHealth()).thenReturn(payload);

        AiServiceHealthException exception = assertThrows(
                AiServiceHealthException.class,
                service::getHealth);

        assertEquals(AiServiceHealthException.Failure.INVALID_RESPONSE, exception.failure());
    }

    @Test
    void rejectsNullPayloadAsInvalidResponse() {
        when(client.getHealth()).thenReturn(null);

        AiServiceHealthException exception = assertThrows(
                AiServiceHealthException.class,
                service::getHealth);

        assertEquals(AiServiceHealthException.Failure.INVALID_RESPONSE, exception.failure());
    }

    @ParameterizedTest
    @MethodSource("clientFailures")
    void preservesClientFailureCategory(
            AiServiceClientException clientException,
            AiServiceHealthException.Failure expectedFailure
    ) {
        when(client.getHealth()).thenThrow(clientException);

        AiServiceHealthException exception = assertThrows(
                AiServiceHealthException.class,
                service::getHealth);

        assertEquals(expectedFailure, exception.failure());
        assertSame(clientException, exception.getCause());
    }

    private static Stream<AiServiceHealthPayload> invalidPayloads() {
        return Stream.of(
                new AiServiceHealthPayload(null, "govbiz-ai-service"),
                new AiServiceHealthPayload("up", null),
                new AiServiceHealthPayload("down", "govbiz-ai-service"),
                new AiServiceHealthPayload("up", "another-service"));
    }

    private static Stream<Arguments> clientFailures() {
        return Stream.of(
                Arguments.of(
                        AiServiceClientException.upstreamError(
                                "unexpected status",
                                new IllegalStateException("test")),
                        AiServiceHealthException.Failure.UPSTREAM_ERROR),
                Arguments.of(
                        AiServiceClientException.invalidResponse(
                                "invalid response",
                                new IllegalArgumentException("test")),
                        AiServiceHealthException.Failure.INVALID_RESPONSE),
                Arguments.of(
                        AiServiceClientException.unavailable(new ConnectException("test")),
                        AiServiceHealthException.Failure.UNAVAILABLE),
                Arguments.of(
                        AiServiceClientException.timeout(new SocketTimeoutException("test")),
                        AiServiceHealthException.Failure.TIMEOUT));
    }
}

File: backend/core-api/src/test/kotlin/ai/govbiz/core/_health_ai_service/service/AiServiceHealthServiceTest.kt
package ai.govbiz.core._health_ai_service.service

import ai.govbiz.core._adapters.ai.client.AiServiceClient
import ai.govbiz.core._adapters.ai.client.AiServiceClientException
import ai.govbiz.core._adapters.ai.client.AiServiceHealthPayload
import java.net.ConnectException
import java.net.SocketTimeoutException
import java.util.stream.Stream
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertSame
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.junit.jupiter.params.ParameterizedTest
import org.junit.jupiter.params.provider.Arguments
import org.junit.jupiter.params.provider.MethodSource
import org.mockito.Mock
import org.mockito.Mockito
import org.mockito.junit.jupiter.MockitoExtension

@ExtendWith(MockitoExtension::class)
class AiServiceHealthServiceTest {

    @Mock
    private lateinit var client: AiServiceClient

    private lateinit var service: AiServiceHealthService

    @BeforeEach
    fun setUp() {
        service = AiServiceHealthService(client)
    }

    @Test
    fun mapsValidatedPayloadToApplicationResult() {
        Mockito.doReturn(AiServiceHealthPayload("up", "govbiz-ai-service"))
            .`when`(client)
            .getHealth()

        val result = service.getHealth()

        assertEquals("up", result.status)
        assertEquals("govbiz-ai-service", result.service)
    }

    @ParameterizedTest
    @MethodSource("invalidPayloads")
    fun rejectsMissingOrUnexpectedLiteralValues(payload: AiServiceHealthPayload) {
        Mockito.doReturn(payload).`when`(client).getHealth()

        val exception = assertThrows(AiServiceHealthException::class.java) {
            service.getHealth()
        }

        assertEquals(AiServiceClientException.Failure.INVALID_RESPONSE, exception.failure)
    }

    @ParameterizedTest
    @MethodSource("clientFailures")
    fun preservesClientFailureCategory(
        clientException: AiServiceClientException,
        expectedFailure: AiServiceClientException.Failure,
    ) {
        Mockito.doThrow(clientException).`when`(client).getHealth()

        val exception = assertThrows(AiServiceHealthException::class.java) {
            service.getHealth()
        }

        assertEquals(expectedFailure, exception.failure)
        assertSame(clientException, exception.cause)
    }

    private companion object {
        @JvmStatic
        fun invalidPayloads(): Stream<AiServiceHealthPayload> =
            Stream.of(
                AiServiceHealthPayload(null, "govbiz-ai-service"),
                AiServiceHealthPayload("up", null),
                AiServiceHealthPayload("down", "govbiz-ai-service"),
                AiServiceHealthPayload("up", "another-service"),
            )

        @JvmStatic
        fun clientFailures(): Stream<Arguments> =
            Stream.of(
                Arguments.of(
                    AiServiceClientException.upstreamError(
                        "unexpected status",
                        IllegalStateException("test"),
                    ),
                    AiServiceClientException.Failure.UPSTREAM_ERROR,
                ),
                Arguments.of(
                    AiServiceClientException.invalidResponse(
                        "invalid response",
                        IllegalArgumentException("test"),
                    ),
                    AiServiceClientException.Failure.INVALID_RESPONSE,
                ),
                Arguments.of(
                    AiServiceClientException.unavailable(ConnectException("test")),
                    AiServiceClientException.Failure.UNAVAILABLE,
                ),
                Arguments.of(
                    AiServiceClientException.timeout(SocketTimeoutException("test")),
                    AiServiceClientException.Failure.TIMEOUT,
                ),
            )
    }
}
